src/common/streak.ts

```diff
@@ -0,0 +1,75 @@
+import { DayOfWeek, DEFAULT_WEEK_START } from './date';
+import {
+  Day,
+  FREEZE_DAYS_IN_A_WEEK,
+  STREAK_RECOVERY_MAX_GAP_DAYS,
+} from './users';
+
+interface DaysProps {
+  day: number;
+  firstDayOfWeek: Day;
+  lastDayOfWeek: Day;
+  lastView: Date;
+}
+
+const getNextDay = (day: number) => (day === 6 ? 0 : day + 1);
+const getPreviousDay = (day: number) => (day === 0 ? 6 : day - 1);
+
+export const getAllowedDays = ({
+  day: dayToday,
+  firstDayOfWeek,
+  lastDayOfWeek,
+  lastView,
+}: DaysProps) => {
+  const lastViewDay = lastView.getDay();
+  const dayTwo = getNextDay(firstDayOfWeek);
+
+  if (
+    lastViewDay === lastDayOfWeek ||
+    dayToday === lastDayOfWeek ||
+    dayToday > dayTwo
+  ) {
+    return STREAK_RECOVERY_MAX_GAP_DAYS;
+  }
+
+  if (dayToday === firstDayOfWeek) {
+    return STREAK_RECOVERY_MAX_GAP_DAYS + FREEZE_DAYS_IN_A_WEEK;
+  }
+
+  const dayOneWeekend = getPreviousDay(lastDayOfWeek);
```

Comment:
what does dayOneWeekend refer to?

Reply:
This is to make it work both on Monday and Sunday as the first weekday. `dayOneWeekend` means Saturday having Monday as the first day of the week and Friday having Sunday as the first day of the week.

Reply:
but the function days getPreviousDay not get first day of week?

Reply:
Let's say the last day of the week is Sunday if Monday is the start of the week. Then the `lastDayOfWeek` (Sunday) minus one is Saturday.

Reply:
Cool then, just this suggestion on my side:

> if it means first day of weekend, let's call it that firstDayOfWeekend or weekendDayOne, it's clearer in my opinion